eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -59,26 +58,39 @@ class Peer(nodeParams: NodeParams, remoteNodeId: PublicKey, authenticator: Actor
   }
 
   when(DISCONNECTED) {
-    case Event(Peer.Connect(NodeURI(_, hostAndPort)), d: DisconnectedData) =>
-      val address = new InetSocketAddress(hostAndPort.getHost, hostAndPort.getPort)
-      if (d.address_opt.contains(address)) {
-        // we already know this address, we'll reconnect automatically
-        sender ! "reconnection in progress"
-        stay
-      } else {
-        // we immediately process explicit connection requests to new addresses
-        context.actorOf(Client.props(nodeParams, authenticator, address, remoteNodeId, origin_opt = Some(sender())))
-        stay
+    case Event(Peer.Connect(_, address_opt), d: DisconnectedData) =>
+      address_opt.map(hAndp => new InetSocketAddress(hAndp.getHost, hAndp.getPort)).orElse {
+        getPeerAddressFromNodeAnnouncement()
+      } match {
+        case None =>
+          log.warning(s"Unable to connect to $remoteNodeId no address found")
+          sender() ! s"Unable to connect to $remoteNodeId no address found"
+          stopPeer()
+        case Some(address) =>
+          if (d.address_opt.contains(address)) { // TODO check if we received a Connect for the same peer but with a different address
+            // we already know this address, we'll reconnect automatically
+            sender ! "reconnection in progress"
+            stay
+          } else {
+            // we immediately process explicit connection requests to new addresses
+            context.actorOf(Client.props(nodeParams, authenticator, address, remoteNodeId, origin_opt = Some(sender())))
+            stay using d.copy(address_opt = Some(address))
+          }
       }
 
     case Event(Reconnect, d: DisconnectedData) =>
-      d.address_opt match {
-        case None => stay // no-op (this peer didn't initiate the connection and doesn't have the ip of the counterparty)
-        case _ if d.channels.isEmpty => stay // no-op (no more channels with this peer)
+      if (d.channels.isEmpty) stay
```

Comment:
It is not obvious at all that the `stay` statement breaks the execution, which is what you seem to be assuming. I find this construction much more error prone than the previous one